# Proyek Analisis Data: [E-Commerce Public Dataset]
- **Nama:** Jeremia Sibarani
- **Email:** sibaranigreamjeremia@gmail.com
- **ID Dicoding:** jeremia_sibarani_12

## Menentukan Pertanyaan Bisnis

- Bagaimana revenue penjualan selama setahun?
- Bagaimana demografi customer dan seller yang dimiliki?
- Product apa yang memiliki skor review terbaik dan terburuk
- Berapa persentase pesanan diantar tepat waktu dan terlambat
- Berapa banyak karyawan yang melunasi pembayaran order
- Kapan terakhir kali pelanggan melakukan transaksi
- Berapa jumlah transaksi pelanggan dalam periode tertentu?
- Berapa total pengeluaran pelanggan pada periode tertentu?

## Import Semua Packages/Library yang Digunakan

In [214]:
import pandas as pd
from tabulate import tabulate
import plotly.express as px
import plotly.express as px
import plotly.graph_objects as go
from plotly.subplots import make_subplots

## Data Wrangling

### Gathering Data

##### Memuat data dengan format csv

In [215]:
customers_df = pd.read_csv('data/customers_dataset.csv')
geolocation_df = pd.read_csv('data/geolocation_dataset.csv')
order_items_df = pd.read_csv('data/order_items_dataset.csv')
order_payments_df = pd.read_csv('data/order_payments_dataset.csv')
order_reviews_df = pd.read_csv('data/order_reviews_dataset.csv')
orders_df = pd.read_csv('data/orders_dataset.csv')
product_category_name_translation_df = pd.read_csv('data/product_category_name_translation.csv')
products_df = pd.read_csv('data/products_dataset.csv')
sellers_df = pd.read_csv('data/sellers_dataset.csv')

In [216]:
customers_df.head()

customer_id                customer_unique_id  \
0  06b8999e2fba1a1fbc88172c00ba8bc7  861eff4711a542e4b93843c6dd7febb0   
1  18955e83d337fd6b2def6b18a428ac77  290c77bc529b7ac935b93aa66c333dc3   
2  4e7b3e00288586ebd08712fdd0374a03  060e732b5b29e8181a18229c7b0b2b5e   
3  b2b6027bc5c5109e529d4dc6358b12c3  259dac757896d24d7702b9acbbff3f3c   
4  4f2d8ab171c80ec8364f7c12e35b23ad  345ecd01c38d18a9036ed96c73b8d066   

   customer_zip_code_prefix          customer_city customer_state  
0                     14409                 franca             SP  
1                      9790  sao bernardo do campo             SP  
2                      1151              sao paulo             SP  
3                      8775        mogi das cruzes             SP  
4                     13056               campinas             SP

In [217]:
geolocation_df.head()

geolocation_zip_code_prefix  geolocation_lat  geolocation_lng  \
0                         1037       -23.545621       -46.639292   
1                         1046       -23.546081       -46.644820   
2                         1046       -23.546129       -46.642951   
3                         1041       -23.544392       -46.639499   
4                         1035       -23.541578       -46.641607   

  geolocation_city geolocation_state  
0        sao paulo                SP  
1        sao paulo                SP  
2        sao paulo                SP  
3        sao paulo                SP  
4        sao paulo                SP

In [218]:
order_items_df.head()

order_id  order_item_id  \
0  00010242fe8c5a6d1ba2dd792cb16214              1   
1  00018f77f2f0320c557190d7a144bdd3              1   
2  000229ec398224ef6ca0657da4fc703e              1   
3  00024acbcdf0a6daa1e931b038114c75              1   
4  00042b26cf59d7ce69dfabb4e55b4fd9              1   

                         product_id                         seller_id  \
0  4244733e06e7ecb4970a6e2683c13e61  48436dade18ac8b2bce089ec2a041202   
1  e5f2d52b802189ee658865ca93d83a8f  dd7ddc04e1b6c2c614352b383efe2d36   
2  c777355d18b72b67abbeef9df44fd0fd  5b51032eddd242adc84c38acab88f23d   
3  7634da152a4610f1595efa32f14722fc  9d7a1d34a5052409006425275ba1c2b4   
4  ac6c3623068f30de03045865e4e10089  df560393f3a51e74553ab94004ba5c87   

   shipping_limit_date   price  freight_value  
0  2017-09-19 09:45:35   58.90          13.29  
1  2017-05-03 11:05:13  239.90          19.93  
2  2018-01-18 14:48:30  199.00          17.87  
3  2018-08-15 10:10:18   12.99          12.79  
4  2017-02-13 13:57:51  199.90          18.14

In [219]:
order_payments_df.head()

order_id  payment_sequential payment_type  \
0  b81ef226f3fe1789b1e8b2acac839d17                   1  credit_card   
1  a9810da82917af2d9aefd1278f1dcfa0                   1  credit_card   
2  25e8ea4e93396b6fa0d3dd708e76c1bd                   1  credit_card   
3  ba78997921bbcdc1373bb41e913ab953                   1  credit_card   
4  42fdf880ba16b47b59251dd489d4441a                   1  credit_card   

   payment_installments  payment_value  
0                     8          99.33  
1                     1          24.39  
2                     1          65.71  
3                     8         107.78  
4                     2         128.45

In [220]:
order_reviews_df.head()

review_id                          order_id  \
0  7bc2406110b926393aa56f80a40eba40  73fc7af87114b39712e6da79b0a377eb   
1  80e641a11e56f04c1ad469d5645fdfde  a548910a1c6147796b98fdf73dbeba33   
2  228ce5500dc1d8e020d8d1322874b6f0  f9e4b658b201a9f2ecdecbb34bed034b   
3  e64fb393e7b32834bb789ff8bb30750e  658677c97b385a9be170737859d3511b   
4  f7c4243c7fe1938f181bec41a392bdeb  8e6bfb81e283fa7e4f11123a3fb894f1   

   review_score review_comment_title  \
0             4                  NaN   
1             5                  NaN   
2             5                  NaN   
3             5                  NaN   
4             5                  NaN   

                              review_comment_message review_creation_date  \
0                                                NaN  2018-01-18 00:00:00   
1                                                NaN  2018-03-10 00:00:00   
2                                                NaN  2018-02-17 00:00:00   
3              Recebi bem antes do prazo estipulado.  2017-04-21 00:00:00   
4  Parabéns lojas lannister adorei comprar pela I...  2018-03-01 00:00:00   

  review_answer_timestamp  
0     2018-01-18 21:46:59  
1     2018-03-11 03:05:13  
2     2018-02-18 14:36:24  
3     2017-04-21 22:02:06  
4     2018-03-02 10:26:53

In [221]:
orders_df.head()

order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1  53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
2  47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   
3  949d5b44dbf5de918fe9c16f97b45f8a  f88197465ea7920adcdbec7375364d82   
4  ad21c59c0840e6cb83a9ceb5573f8159  8ab97904e6daea8866dbdbc4fb7aad2c   

  order_status order_purchase_timestamp    order_approved_at  \
0    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
1    delivered      2018-07-24 20:41:37  2018-07-26 03:24:27   
2    delivered      2018-08-08 08:38:49  2018-08-08 08:55:23   
3    delivered      2017-11-18 19:28:06  2017-11-18 19:45:59   
4    delivered      2018-02-13 21:18:39  2018-02-13 22:20:29   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-10-04 19:55:00           2017-10-10 21:25:13   
1          2018-07-26 14:31:00           2018-08-07 15:27:45   
2          2018-08-08 13:50:00           2018-08-17 18:06:29   
3          2017-11-22 13:39:59           2017-12-02 00:28:42   
4          2018-02-14 19:46:34           2018-02-16 18:17:02   

  order_estimated_delivery_date  
0           2017-10-18 00:00:00  
1           2018-08-13 00:00:00  
2           2018-09-04 00:00:00  
3           2017-12-15 00:00:00  
4           2018-02-26 00:00:00

In [222]:
products_df.head()

product_id  product_category_name  \
0  1e9e8ef04dbcff4541ed26657ea517e5             perfumaria   
1  3aa071139cb16b67ca9e5dea641aaa2f                  artes   
2  96bd76ec8810374ed1b65e291975717f          esporte_lazer   
3  cef67bcfe19066a932b7673e239eb23d                  bebes   
4  9dc1a7de274444849c219cff195d0b71  utilidades_domesticas   

   product_name_lenght  product_description_lenght  product_photos_qty  \
0                 40.0                       287.0                 1.0   
1                 44.0                       276.0                 1.0   
2                 46.0                       250.0                 1.0   
3                 27.0                       261.0                 1.0   
4                 37.0                       402.0                 4.0   

   product_weight_g  product_length_cm  product_height_cm  product_width_cm  
0             225.0               16.0               10.0              14.0  
1            1000.0               30.0               18.0              20.0  
2             154.0               18.0                9.0              15.0  
3             371.0               26.0                4.0              26.0  
4             625.0               20.0               17.0              13.0

In [223]:
product_category_name_translation_df.head()

product_category_name product_category_name_english
0            beleza_saude                 health_beauty
1  informatica_acessorios         computers_accessories
2              automotivo                          auto
3         cama_mesa_banho                bed_bath_table
4        moveis_decoracao               furniture_decor

In [224]:
sellers_df.head()

seller_id  seller_zip_code_prefix  \
0  3442f8959a84dea7ee197c632cb2df15                   13023   
1  d1b65fc7debc3361ea86b5f14c68d2e2                   13844   
2  ce3ad9de960102d0677a81f5d0bb7b2d                   20031   
3  c0f3eea2e14555b6faeea3dd58c1b1c3                    4195   
4  51a04a8a6bdcb23deccc82b0b80742cf                   12914   

         seller_city seller_state  
0           campinas           SP  
1         mogi guacu           SP  
2     rio de janeiro           RJ  
3          sao paulo           SP  
4  braganca paulista           SP

**Insight:**
- Terdapat total 7 data yang digunakan, yaitu Customers, Geolocations, Orders, OrderPayments, OrderReviews, Products, Sellers
- Dengan melihat 5 baris teratas dari setiap data, terdapat baris yang memiliki nilai NaN, yaitu pada tabel OrderReviews di kolom review_comment_title dan review_comment_message
- Terdapat beberapa kolom yang sepertinya memiliki hubungan berdasarkan konsep relasi, seperti pada data Orders, terdapat kolom product_id dan seller_id yang mungkin masing - masing digunakan sebagai dasar relasi dengan data Products dan Sellers, hal ini mungkin dapat ditelusuri lebih lanjut di tahap Assesing data

### Assessing Data

##### Membuat fungsi untuk melihat laporan nilai null dan NaN dari sebuah dataframe

In [225]:
def get_df_report(df):
  res = {}
  for (col1, nullvalue), (_, navalue), (_, totalvalue), (_, duplicatevalue) in zip(df.isnull().sum().items(), df.isna().sum().items(), df.count().items(), df.duplicated().items()):
    report = {}
    report['null'] = nullvalue
    report['na'] = navalue
    report['duplicate'] = duplicatevalue
    report['null_percentage'] = round(nullvalue / (totalvalue + nullvalue + navalue) * 100, 2)
    report['na_percentage'] = round(navalue / (totalvalue + nullvalue + navalue) * 100, 2)
    report['duplicate_percentage'] = round(duplicatevalue / (totalvalue + nullvalue + navalue) * 100, 2)
    report['total'] = (totalvalue + nullvalue + navalue)
    res[col1] = report
  return pd.DataFrame.from_dict(res, orient='index')

##### Missing value dan duplicate value <br>
Melihat laporan missing value dan duplicate value

In [226]:
print(tabulate(get_df_report(customers_df), headers='keys', tablefmt='psql'))
print(tabulate(get_df_report(geolocation_df), headers='keys', tablefmt='psql'))
print(tabulate(get_df_report(order_items_df), headers='keys', tablefmt='psql'))
print(tabulate(get_df_report(orders_df), headers='keys', tablefmt='psql'))
print(tabulate(get_df_report(order_payments_df), headers='keys', tablefmt='psql'))
print(tabulate(get_df_report(order_reviews_df), headers='keys', tablefmt='psql'))
print(tabulate(get_df_report(products_df), headers='keys', tablefmt='psql'))
print(tabulate(get_df_report(product_category_name_translation_df), headers='keys', tablefmt='psql'))
print(tabulate(get_df_report(sellers_df), headers='keys', tablefmt='psql'))

+--------------------------+--------+------+-------------+-------------------+-----------------+------------------------+---------+
|                          |   null |   na | duplicate   |   null_percentage |   na_percentage |   duplicate_percentage |   total |
|--------------------------+--------+------+-------------+-------------------+-----------------+------------------------+---------|
| customer_id              |      0 |    0 | False       |                 0 |               0 |                      0 |   99441 |
| customer_unique_id       |      0 |    0 | False       |                 0 |               0 |                      0 |   99441 |
| customer_zip_code_prefix |      0 |    0 | False       |                 0 |               0 |                      0 |   99441 |
| customer_city            |      0 |    0 | False       |                 0 |               0 |                      0 |   99441 |
| customer_state           |      0 |    0 | False       |                 0

In [227]:
customers_df.describe(include='all')

customer_id                customer_unique_id  \
count                              99441                             99441   
unique                             99441                             96096   
top     06b8999e2fba1a1fbc88172c00ba8bc7  8d50f5eadf50201ccdcedfb9e2ac8455   
freq                                   1                                17   
mean                                 NaN                               NaN   
std                                  NaN                               NaN   
min                                  NaN                               NaN   
25%                                  NaN                               NaN   
50%                                  NaN                               NaN   
75%                                  NaN                               NaN   
max                                  NaN                               NaN   

        customer_zip_code_prefix customer_city customer_state  
count               99441.000000         99441          99441  
unique                       NaN          4119             27  
top                          NaN     sao paulo             SP  
freq                         NaN         15540          41746  
mean                35137.474583           NaN            NaN  
std                 29797.938996           NaN            NaN  
min                  1003.000000           NaN            NaN  
25%                 11347.000000           NaN            NaN  
50%                 24416.000000           NaN            NaN  
75%                 58900.000000           NaN            NaN  
max                 99990.000000           NaN            NaN

In [228]:
geolocation_df.describe(include='all')

geolocation_zip_code_prefix  geolocation_lat  geolocation_lng  \
count                  1.000163e+06     1.000163e+06     1.000163e+06   
unique                          NaN              NaN              NaN   
top                             NaN              NaN              NaN   
freq                            NaN              NaN              NaN   
mean                   3.657417e+04    -2.117615e+01    -4.639054e+01   
std                    3.054934e+04     5.715866e+00     4.269748e+00   
min                    1.001000e+03    -3.660537e+01    -1.014668e+02   
25%                    1.107500e+04    -2.360355e+01    -4.857317e+01   
50%                    2.653000e+04    -2.291938e+01    -4.663788e+01   
75%                    6.350400e+04    -1.997962e+01    -4.376771e+01   
max                    9.999000e+04     4.506593e+01     1.211054e+02   

       geolocation_city geolocation_state  
count           1000163           1000163  
unique             8011                27  
top           sao paulo                SP  
freq             135800            404268  
mean                NaN               NaN  
std                 NaN               NaN  
min                 NaN               NaN  
25%                 NaN               NaN  
50%                 NaN               NaN  
75%                 NaN               NaN  
max                 NaN               NaN

In [229]:
order_items_df.describe(include='all')

order_id  order_item_id  \
count                             112650  112650.000000   
unique                             98666            NaN   
top     8272b63d03f5f79c56e9e4120aec44ef            NaN   
freq                                  21            NaN   
mean                                 NaN       1.197834   
std                                  NaN       0.705124   
min                                  NaN       1.000000   
25%                                  NaN       1.000000   
50%                                  NaN       1.000000   
75%                                  NaN       1.000000   
max                                  NaN      21.000000   

                              product_id                         seller_id  \
count                             112650                            112650   
unique                             32951                              3095   
top     aca2eb7d00ea1a7b8ebd4e68314663af  6560211a19b47992c3666cc44a7e94c0   
freq                                 527                              2033   
mean                                 NaN                               NaN   
std                                  NaN                               NaN   
min                                  NaN                               NaN   
25%                                  NaN                               NaN   
50%                                  NaN                               NaN   
75%                                  NaN                               NaN   
max                                  NaN                               NaN   

        shipping_limit_date          price  freight_value  
count                112650  112650.000000  112650.000000  
unique                93318            NaN            NaN  
top     2017-07-21 18:25:23            NaN            NaN  
freq                     21            NaN            NaN  
mean                    NaN     120.653739      19.990320  
std                     NaN     183.633928      15.806405  
min                     NaN       0.850000       0.000000  
25%                     NaN      39.900000      13.080000  
50%                     NaN      74.990000      16.260000  
75%                     NaN     134.900000      21.150000  
max                     NaN    6735.000000     409.680000

In [230]:
orders_df.describe(include='all')

order_id                       customer_id  \
count                              99441                             99441   
unique                             99441                             99441   
top     e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
freq                                   1                                 1   

       order_status order_purchase_timestamp    order_approved_at  \
count         99441                    99441                99281   
unique            8                    98875                90733   
top       delivered      2018-04-11 10:48:14  2018-02-27 04:31:10   
freq          96478                        3                    9   

       order_delivered_carrier_date order_delivered_customer_date  \
count                         97658                         96476   
unique                        81018                         95664   
top             2018-05-09 15:48:00           2018-05-08 23:38:46   
freq                             47                             3   

       order_estimated_delivery_date  
count                          99441  
unique                           459  
top              2017-12-20 00:00:00  
freq                             522

In [231]:
order_reviews_df.describe(include='all')

review_id                          order_id  \
count                              99224                             99224   
unique                             98410                             98673   
top     7b606b0d57b078384f0b58eac1d41d78  c88b1d1b157a9999ce368f218a407141   
freq                                   3                                 3   
mean                                 NaN                               NaN   
std                                  NaN                               NaN   
min                                  NaN                               NaN   
25%                                  NaN                               NaN   
50%                                  NaN                               NaN   
75%                                  NaN                               NaN   
max                                  NaN                               NaN   

        review_score review_comment_title review_comment_message  \
count   99224.000000                11568                  40977   
unique           NaN                 4527                  36159   
top              NaN            Recomendo              Muito bom   
freq             NaN                  423                    230   
mean        4.086421                  NaN                    NaN   
std         1.347579                  NaN                    NaN   
min         1.000000                  NaN                    NaN   
25%         4.000000                  NaN                    NaN   
50%         5.000000                  NaN                    NaN   
75%         5.000000                  NaN                    NaN   
max         5.000000                  NaN                    NaN   

       review_creation_date review_answer_timestamp  
count                 99224                   99224  
unique                  636                   98248  
top     2017-12-19 00:00:00     2017-06-15 23:21:05  
freq                    463                       4  
mean                    NaN                     NaN  
std                     NaN                     NaN  
min                     NaN                     NaN  
25%                     NaN                     NaN  
50%                     NaN                     NaN  
75%                     NaN                     NaN  
max                     NaN                     NaN

In [232]:
order_payments_df.describe(include='all')

order_id  payment_sequential payment_type  \
count                             103886       103886.000000       103886   
unique                             99440                 NaN            5   
top     fa65dad1b0e818e3ccc5cb0e39231352                 NaN  credit_card   
freq                                  29                 NaN        76795   
mean                                 NaN            1.092679          NaN   
std                                  NaN            0.706584          NaN   
min                                  NaN            1.000000          NaN   
25%                                  NaN            1.000000          NaN   
50%                                  NaN            1.000000          NaN   
75%                                  NaN            1.000000          NaN   
max                                  NaN           29.000000          NaN   

        payment_installments  payment_value  
count          103886.000000  103886.000000  
unique                   NaN            NaN  
top                      NaN            NaN  
freq                     NaN            NaN  
mean                2.853349     154.100380  
std                 2.687051     217.494064  
min                 0.000000       0.000000  
25%                 1.000000      56.790000  
50%                 1.000000     100.000000  
75%                 4.000000     171.837500  
max                24.000000   13664.080000

##### Melihat data payment
Terdapat informasi payment installment, payment value, dan payment sequential, dengan keterangan sebagai berikut : <br>
- payment_installments : cicilan
- payment_value : nilai pembayaran
- payment_sequential : urutan pembayaran, jika menyicil akan terdapat urutan 1,2,3,4, dst, dan masing - masing memiliki order_id yang sama, sementara jika tidak menyicil hanya akan urutan 1 dan tidak ada order_id yang sama di baris lain

In [233]:
order_payments_df[(order_payments_df['payment_installments']) == 0 | (order_payments_df['payment_value'] == 0)]

order_id  payment_sequential payment_type  \
19922   8bcbe01d44d147f901cd3192671144db                   4      voucher   
36822   fa65dad1b0e818e3ccc5cb0e39231352                  14      voucher   
43744   6ccb433e00daae1283ccc956189c82ae                   4      voucher   
46982   744bade1fcf9ff3f31d860ace076d422                   2  credit_card   
51280   4637ca194b6387e2d538dc89b124b0ee                   1  not_defined   
57411   00b1cb0320190ca0daa2c88b35206009                   1  not_defined   
62674   45ed6e85398a87c253db47c2d9f48216                   3      voucher   
77885   fa65dad1b0e818e3ccc5cb0e39231352                  13      voucher   
79014   1a57108394169c0b47d8f876acc9ba2d                   2  credit_card   
94427   c8c528189310eaa44a745b8d9d26908b                   1  not_defined   
100766  b23878b3e8eb4d25a158f57d96331b18                   4      voucher   

        payment_installments  payment_value  
19922                      1           0.00  
36822                      1           0.00  
43744                      1           0.00  
46982                      0          58.69  
51280                      1           0.00  
57411                      1           0.00  
62674                      1           0.00  
77885                      1           0.00  
79014                      0         129.94  
94427                      1           0.00  
100766                     1           0.00

Berdasarkan tabel OrderPayments di atas, dapat dilihat beberapa ketidakakuratan data, seperti :
- Nilai payment_installments 0, hal ini tidak mungkin terjadi karena nilai ini setidaknya harus 1, yang artinya membayar secara lunas dengan tidak menyicil, jika bernilai 0, maka pembayaran tidak pernah dilakukan namun payment_value tidak 0, 2 hal ini berkontradiksi
- Nilai payment_value 0, padahal nilai payment_sequential dan payment_installments tidak 0, hal ini mungkin saja berarti bahwa pada saat pembayaran, transaksi gagal atau ditolak. namun data pembayaran masih terekam

In [234]:
products_df.describe(include='all')

product_id product_category_name  \
count                              32951                 32341   
unique                             32951                    73   
top     1e9e8ef04dbcff4541ed26657ea517e5       cama_mesa_banho   
freq                                   1                  3029   
mean                                 NaN                   NaN   
std                                  NaN                   NaN   
min                                  NaN                   NaN   
25%                                  NaN                   NaN   
50%                                  NaN                   NaN   
75%                                  NaN                   NaN   
max                                  NaN                   NaN   

        product_name_lenght  product_description_lenght  product_photos_qty  \
count          32341.000000                32341.000000        32341.000000   
unique                  NaN                         NaN                 NaN   
top                     NaN                         NaN                 NaN   
freq                    NaN                         NaN                 NaN   
mean              48.476949                  771.495285            2.188986   
std               10.245741                  635.115225            1.736766   
min                5.000000                    4.000000            1.000000   
25%               42.000000                  339.000000            1.000000   
50%               51.000000                  595.000000            1.000000   
75%               57.000000                  972.000000            3.000000   
max               76.000000                 3992.000000           20.000000   

        product_weight_g  product_length_cm  product_height_cm  \
count       32949.000000       32949.000000       32949.000000   
unique               NaN                NaN                NaN   
top                  NaN                NaN                NaN   
freq                 NaN                NaN                NaN   
mean         2276.472488          30.815078          16.937661   
std          4282.038731          16.914458          13.637554   
min             0.000000           7.000000           2.000000   
25%           300.000000          18.000000           8.000000   
50%           700.000000          25.000000          13.000000   
75%          1900.000000          38.000000          21.000000   
max         40425.000000         105.000000         105.000000   

        product_width_cm  
count       32949.000000  
unique               NaN  
top                  NaN  
freq                 NaN  
mean           23.196728  
std            12.079047  
min             6.000000  
25%            15.000000  
50%            20.000000  
75%            30.000000  
max           118.000000

In [235]:
sellers_df.describe(include='all')

seller_id  seller_zip_code_prefix seller_city  \
count                               3095             3095.000000        3095   
unique                              3095                     NaN         611   
top     3442f8959a84dea7ee197c632cb2df15                     NaN   sao paulo   
freq                                   1                     NaN         694   
mean                                 NaN            32291.059451         NaN   
std                                  NaN            32713.453830         NaN   
min                                  NaN             1001.000000         NaN   
25%                                  NaN             7093.500000         NaN   
50%                                  NaN            14940.000000         NaN   
75%                                  NaN            64552.500000         NaN   
max                                  NaN            99730.000000         NaN   

       seller_state  
count          3095  
unique           23  
top              SP  
freq           1849  
mean            NaN  
std             NaN  
min             NaN  
25%             NaN  
50%             NaN  
75%             NaN  
max             NaN

**Insight:**
- Kota dengan customer terbanyak adalah Sao Paulo, state SP, yaitu sebanyak 15540 atau sebesar 15.6%
- Terdapat transaksi bernilai 0 pada data OrderPayments, dan terdapat nilai payment installments (cicilan) bernilai 0, hal ini bisa mengindikasikan pembayaran cicilan yang sudah lunas atau memang transaksi tersebut tidak pernah ada
- Terdapat product dengan berat 0 gram, hal ini mungkin menjadi salah satu indikasi inaccurate data, karena tidak mungkin ada suatu benda yang tidak memiliki berat
- Terdapat 3 tabel yang memiliki missing value, yaitu OrderReviews, Products, dan Orders berdasarkan prinsip data cleaning terdapat beberapa metode untuk mengatasi missing value, yaitu : Dropping, Imputation, Interpolation
- Data OrderReviews tidak cocok diterapkan dropping karena persentase missing values cukup besar, yaitu pada rentang 36.99% sampai 46.90%, kemungkinan diterapkan teknik imputation, karena metode interpolation hanya cocok digunakan untuk data numerik
- Data Products dan Orders cocok diterapkan teknik dropping karena persentase missing values relative kecil, yaitu pada rentang 0.01% sampai 1.82% dan pada rentang 0.16% sampai 2.9% masing - masing untuk data Products dan Orders
- Tidak terdapat duplicate value pada data secara keseluruhan

### Cleaning Data

#### Dropping <br>
Berdasarkan uraian pada tahap assesing, untuk mengatasi missing value pada data Products digunakan teknik dropping

In [236]:
products_df.dropna(axis=0,inplace=True)
products_df.isnull().sum()

product_id                    0
product_category_name         0
product_name_lenght           0
product_description_lenght    0
product_photos_qty            0
product_weight_g              0
product_length_cm             0
product_height_cm             0
product_width_cm              0
dtype: int64

In [237]:
orders_df.dropna(axis=0, inplace=True)
orders_df.isnull().sum()

order_id                         0
customer_id                      0
order_status                     0
order_purchase_timestamp         0
order_approved_at                0
order_delivered_carrier_date     0
order_delivered_customer_date    0
order_estimated_delivery_date    0
dtype: int64

#### Imputation <br>
Pada data OrderReviews, teknik imputation digunakan untuk mengganti missing values dengan nilai tertentu. Metode ini lebih dipilih dibandingkan interpolasi karena missing values berupa data kategorikal. Namun, kolom review_comment_title dan review_comment_message berisi informasi acak yang tidak dapat dikelompokkan. Jika kasusnya adalah gender, missing values dapat digantikan dengan label gender yang paling sering muncul. Namun, karena hal ini tidak berlaku untuk komentar review, missing values diganti dengan "Tidak ada review" dengan asumsi bahwa nilai null berarti pengguna tidak memberikan review.

In [238]:
order_reviews_df.fillna('Tidak ada review', inplace=True)
order_reviews_df.isnull().sum()

review_id                  0
order_id                   0
review_score               0
review_comment_title       0
review_comment_message     0
review_creation_date       0
review_answer_timestamp    0
dtype: int64

#### Inaccurate data <br>
- OrderPayments <br>
Seperti yang telah dijelaskan bahwa beberapa data pada tabel OrderPayments mengalami ketidakakuratan pada kolom payment_installments dan payment_value. Untuk mengatasi hal ini kita dapat melakukan dropping pada baris yang memiliki kondisi ini

In [239]:
order_payments_df.drop(
  order_payments_df[(order_payments_df['payment_installments']) == 0 | (order_payments_df['payment_value'] == 0)].index,
  inplace=True
)
order_payments_df.head()

order_id  payment_sequential payment_type  \
0  b81ef226f3fe1789b1e8b2acac839d17                   1  credit_card   
1  a9810da82917af2d9aefd1278f1dcfa0                   1  credit_card   
2  25e8ea4e93396b6fa0d3dd708e76c1bd                   1  credit_card   
3  ba78997921bbcdc1373bb41e913ab953                   1  credit_card   
4  42fdf880ba16b47b59251dd489d4441a                   1  credit_card   

   payment_installments  payment_value  
0                     8          99.33  
1                     1          24.39  
2                     1          65.71  
3                     8         107.78  
4                     2         128.45

- Products <br>
Terdapat ketidakakuratan data pada kolom product_weight pada data products, pada masalah ini juga diterapkan teknik droping

In [240]:
products_df.drop(
  products_df[products_df['product_weight_g'] == 0].index,
  inplace=True
)
products_df.head()

product_id  product_category_name  \
0  1e9e8ef04dbcff4541ed26657ea517e5             perfumaria   
1  3aa071139cb16b67ca9e5dea641aaa2f                  artes   
2  96bd76ec8810374ed1b65e291975717f          esporte_lazer   
3  cef67bcfe19066a932b7673e239eb23d                  bebes   
4  9dc1a7de274444849c219cff195d0b71  utilidades_domesticas   

   product_name_lenght  product_description_lenght  product_photos_qty  \
0                 40.0                       287.0                 1.0   
1                 44.0                       276.0                 1.0   
2                 46.0                       250.0                 1.0   
3                 27.0                       261.0                 1.0   
4                 37.0                       402.0                 4.0   

   product_weight_g  product_length_cm  product_height_cm  product_width_cm  
0             225.0               16.0               10.0              14.0  
1            1000.0               30.0               18.0              20.0  
2             154.0               18.0                9.0              15.0  
3             371.0               26.0                4.0              26.0  
4             625.0               20.0               17.0              13.0

**Insight:**
- Penerapan teknik droping memangkas sebesar 0.01% sampai 1.82% baris pada data OrderReviews
- Penerapan teknik imputation mengganti nilai null menjadi "Tidak ada review" dengan asumsi bahwa nilai null berarti tidak memberikan review
- Penerapan teknik droping pada data OrderPayments dan Products mengeliminasi data yang tidak akurat
- Data yang diperoleh sudah relative bersih

## Exploratory Data Analysis (EDA)

### Explore ...

Kondisi revenue data berdasarkan gabungan tabel order dan order items

In [241]:
order_order_id =  pd.merge(
  left=orders_df,
  right=order_items_df,
  left_on="order_id",
  right_on="order_id"
)

order_order_id['purchase_year'] = pd.to_datetime(order_order_id['order_purchase_timestamp']).dt.year
order_order_id['purchase_month'] = pd.to_datetime(order_order_id['order_purchase_timestamp']).dt.month
order_order_id['purchase_day'] = pd.to_datetime(order_order_id['order_purchase_timestamp']).dt.day
order_order_id.head()

order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1  53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
2  47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   
3  949d5b44dbf5de918fe9c16f97b45f8a  f88197465ea7920adcdbec7375364d82   
4  ad21c59c0840e6cb83a9ceb5573f8159  8ab97904e6daea8866dbdbc4fb7aad2c   

  order_status order_purchase_timestamp    order_approved_at  \
0    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
1    delivered      2018-07-24 20:41:37  2018-07-26 03:24:27   
2    delivered      2018-08-08 08:38:49  2018-08-08 08:55:23   
3    delivered      2017-11-18 19:28:06  2017-11-18 19:45:59   
4    delivered      2018-02-13 21:18:39  2018-02-13 22:20:29   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-10-04 19:55:00           2017-10-10 21:25:13   
1          2018-07-26 14:31:00           2018-08-07 15:27:45   
2          2018-08-08 13:50:00           2018-08-17 18:06:29   
3          2017-11-22 13:39:59           2017-12-02 00:28:42   
4          2018-02-14 19:46:34           2018-02-16 18:17:02   

  order_estimated_delivery_date  order_item_id  \
0           2017-10-18 00:00:00              1   
1           2018-08-13 00:00:00              1   
2           2018-09-04 00:00:00              1   
3           2017-12-15 00:00:00              1   
4           2018-02-26 00:00:00              1   

                         product_id                         seller_id  \
0  87285b34884572647811a353c7ac498a  3504c0cb71d7fa48d967e0e4c94d59d9   
1  595fac2a385ac33a80bd5114aec74eb8  289cdb325fb7e7f891c38608bf9e0962   
2  aa4383b373c6aca5d8797843e5594415  4869f7a5dfa277a7dca6462dcf3b52b2   
3  d0b61bfb1de832b15ba9d266ca96e5b0  66922902710d126a0e7d26b0e3805106   
4  65266b2da20d04dbe00c5c2d3bb7859e  2c9e548be18521d1c43cde1c582c6de8   

   shipping_limit_date   price  freight_value  purchase_year  purchase_month  \
0  2017-10-06 11:07:15   29.99           8.72           2017              10   
1  2018-07-30 03:24:27  118.70          22.76           2018               7   
2  2018-08-13 08:55:23  159.90          19.22           2018               8   
3  2017-11-23 19:45:59   45.00          27.20           2017              11   
4  2018-02-19 20:31:37   19.90           8.72           2018               2   

   purchase_day  
0             2  
1            24  
2             8  
3            18  
4            13

#### Total harga order selama beberapa bulan dalam beberapa tahun

In [242]:
order_order_id.groupby(by=['purchase_month', 'purchase_year', 'purchase_day']).agg({
  "price": "sum"
})

price
purchase_month purchase_year purchase_day          
1              2017          5               396.90
                             6               916.38
                             7              1351.90
                             8               449.78
                             9               673.79
...                                             ...
12             2017          27            17341.40
                             28            14647.79
                             29            15912.82
                             30            13042.65
                             31             8462.88

[612 rows x 1 columns]

#### Sebaran customer berdasarkan negara dan kota

In [243]:
customer_by_state_city = customers_df.groupby(by=["customer_state", "customer_city"]).agg({
  "customer_id" : "nunique"
}).sort_values(ascending=False, by='customer_id')

#### 10 kota dengan customer paling banyak

In [244]:
customer_by_state_city.head(10)

customer_id
customer_state customer_city                     
SP             sao paulo                    15540
RJ             rio de janeiro                6882
MG             belo horizonte                2773
DF             brasilia                      2131
PR             curitiba                      1521
SP             campinas                      1444
RS             porto alegre                  1379
BA             salvador                      1245
SP             guarulhos                     1189
               sao bernardo do campo          938

#### 10 Kota dengan customer paling sedikit

In [245]:
customer_by_state_city.tail(10)

customer_id
customer_state customer_city             
RS             colinas                  1
               ciriaco                  1
MG             ilicinea                 1
RS             chiapetta                1
MG             ingai                    1
RS             cerro grande             1
               cerrito                  1
               caseiros                 1
MG             ipuiuna                  1
AC             brasileia                1

#### Melihat tanggapan pelanggan terhadap barang yang dibeli

Tipe rating

In [246]:
order_reviews_df.groupby(by="review_score").agg({
  "order_id": "nunique"
})

order_id
review_score          
1                11393
2                 3148
3                 8160
4                19098
5                57076

#### Data order dengan review. <br> 
Sesuai Entity Relational Diagram, beberapa table seperti : order item, product, order review, dan order harus digabungkan melalui fungsi merge untuk memperoleh data order dan review

In [247]:
order_items_products = pd.merge(
  left=order_items_df,
  right=products_df,
  left_on="product_id",
  right_on="product_id"
)

orders_order_items_products = pd.merge(
  left=orders_df,
  right=order_items_products,
  left_on="order_id",
  right_on="order_id",
  how="left"
)

orders_products_review = pd.merge(
  left=orders_order_items_products,
  right=order_reviews_df,
  left_on="order_id",
  right_on="order_id",
  how="left"
)
orders_products_review.head()

order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1  53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
2  47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   
3  949d5b44dbf5de918fe9c16f97b45f8a  f88197465ea7920adcdbec7375364d82   
4  ad21c59c0840e6cb83a9ceb5573f8159  8ab97904e6daea8866dbdbc4fb7aad2c   

  order_status order_purchase_timestamp    order_approved_at  \
0    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
1    delivered      2018-07-24 20:41:37  2018-07-26 03:24:27   
2    delivered      2018-08-08 08:38:49  2018-08-08 08:55:23   
3    delivered      2017-11-18 19:28:06  2017-11-18 19:45:59   
4    delivered      2018-02-13 21:18:39  2018-02-13 22:20:29   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-10-04 19:55:00           2017-10-10 21:25:13   
1          2018-07-26 14:31:00           2018-08-07 15:27:45   
2          2018-08-08 13:50:00           2018-08-17 18:06:29   
3          2017-11-22 13:39:59           2017-12-02 00:28:42   
4          2018-02-14 19:46:34           2018-02-16 18:17:02   

  order_estimated_delivery_date  order_item_id  \
0           2017-10-18 00:00:00            1.0   
1           2018-08-13 00:00:00            1.0   
2           2018-09-04 00:00:00            1.0   
3           2017-12-15 00:00:00            1.0   
4           2018-02-26 00:00:00            1.0   

                         product_id  ... product_weight_g product_length_cm  \
0  87285b34884572647811a353c7ac498a  ...            500.0              19.0   
1  595fac2a385ac33a80bd5114aec74eb8  ...            400.0              19.0   
2  aa4383b373c6aca5d8797843e5594415  ...            420.0              24.0   
3  d0b61bfb1de832b15ba9d266ca96e5b0  ...            450.0              30.0   
4  65266b2da20d04dbe00c5c2d3bb7859e  ...            250.0              51.0   

   product_height_cm  product_width_cm                         review_id  \
0                8.0              13.0  a54f0611adc9ed256b57ede6b6eb5114   
1               13.0              19.0  8d5266042046a06655c8db133d120ba5   
2               19.0              21.0  e73b67b67587f7644d5bd1a52deb1b01   
3               10.0              20.0  359d03e676b3c069f62cadba8dd3f6e8   
4               15.0              15.0  e50934924e227544ba8246aeb3770dd4   

   review_score  review_comment_title  \
0           4.0      Tidak ada review   
1           4.0      Muito boa a loja   
2           5.0      Tidak ada review   
3           5.0      Tidak ada review   
4           5.0      Tidak ada review   

                              review_comment_message  review_creation_date  \
0  Não testei o produto ainda, mas ele veio corre...   2017-10-11 00:00:00   
1                               Muito bom o produto.   2018-08-08 00:00:00   
2                                   Tidak ada review   2018-08-18 00:00:00   
3  O produto foi exatamente o que eu esperava e e...   2017-12-03 00:00:00   
4                                   Tidak ada review   2018-02-17 00:00:00   

   review_answer_timestamp  
0      2017-10-12 03:43:48  
1      2018-08-08 18:37:50  
2      2018-08-22 19:07:58  
3      2017-12-05 19:21:58  
4      2018-02-18 13:02:51  

[5 rows x 28 columns]

#### Total skor orderan berdasarkan nama product

In [248]:
product_by_review = orders_products_review.groupby(by="product_category_name").agg({
  "review_score": "sum"
}).sort_values(ascending=False, by="review_score")

#### 10 Product dengan skor review paling tinggi

In [249]:
product_by_review.head(10)

review_score
product_category_name               
cama_mesa_banho              43037.0
beleza_saude                 39636.0
esporte_lazer                35135.0
moveis_decoracao             32212.0
informatica_acessorios       30570.0
utilidades_domesticas        27855.0
relogios_presentes           23709.0
telefonia                    17602.0
ferramentas_jardim           17357.0
automotivo                   16942.0

#### 10 Product dengan skor review paling rendah

In [250]:
product_by_review.tail(10)

review_score
product_category_name                                      
fashion_esporte                                       131.0
fraldas_higiene                                       125.0
artes_e_artesanato                                     99.0
casa_conforto_2                                        98.0
cds_dvds_musicais                                      65.0
la_cuisine                                             52.0
portateis_cozinha_e_preparadores_de_alimentos          48.0
fashion_roupa_infanto_juvenil                          35.0
pc_gamer                                               29.0
seguros_e_servicos                                      5.0

#### Tipe orderan <br>
Banyaknya orderan dengan status delivered yang tiba tepat waktu, akan dibagi menjadi 3 kategori : <br>
- on time : order_status delivered dan tepat waktu
- late : order_status delivered dan tidak tepat waktu
- other : order_status bukan delivered

In [251]:
def filter_delivery_order(delivered_customer_date, estimated_delivery_date, delivery_status):
  if delivery_status == 'delivered':
    if delivered_customer_date <= estimated_delivery_date:
      return "on time"
    else:
      return "late"
  else:
    return "other"

orders_df['delivery_status'] = orders_df.apply(lambda x: filter_delivery_order(x.order_delivered_customer_date, x.order_estimated_delivery_date, x.order_status), axis=1)

In [252]:
orders_df.groupby(by='delivery_status').order_id.nunique().sort_values(ascending=False)

delivery_status
on time    88630
late        7825
other          6
Name: order_id, dtype: int64

#### Kondisi pembayaran pelanggan terhadap orderan. <br>
Total tagihan pelanggan terhadap suatu product meliputi harga barang dan biaya pengiriman. Pelanggan diperbolehkan untuk melakukan penyicilan pada saat melakukan pembayaran. Pelunasan tagihan oleh pelanggan dapat dirumuskan seperti berikut : <br>

$$ p = b + f $$ <br>
$$ c = p - (a_1 + a_2 + a_3 + ... + a_n) $$

$$ if~c < 1~lunas, else~tidak~lunas $$ <br>
Ket :
- n : banyak cicilan
- p : tagihan pelanggan
- b : harga barang
- f : biaya pengiriman
- a : cicilan
- c : selisih pembayaran dengan tagihan

Total pembayaran pelanggan berdasarkan customer id

In [253]:
payment_by_order_id = order_payments_df.groupby(by="order_id").agg({
    "payment_value" : "sum"
})
payment_by_order_id.head()

payment_value
order_id                                       
00010242fe8c5a6d1ba2dd792cb16214          72.19
00018f77f2f0320c557190d7a144bdd3         259.83
000229ec398224ef6ca0657da4fc703e         216.87
00024acbcdf0a6daa1e931b038114c75          25.78
00042b26cf59d7ce69dfabb4e55b4fd9         218.04

Untuk memperoleh informasi pembayaran dan customer, maka tabel order, order payment, dan customer harus dibabungkan seperti berikut :

In [254]:
orders_items = pd.merge(
    left=orders_df,
    right=order_items_df,
    left_on="order_id",
    right_on="order_id"
)


orders_order_payments = pd.merge(
    left=orders_items,
    right=order_payments_df,
    left_on="order_id",
    right_on="order_id"
)

customer_payment = pd.merge(
    left=orders_order_payments,
    right=customers_df,
    left_on="customer_id",
    right_on="customer_id"
)

customer_payment.head()

order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
2  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
3  53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
4  47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   

  order_status order_purchase_timestamp    order_approved_at  \
0    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
1    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
2    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
3    delivered      2018-07-24 20:41:37  2018-07-26 03:24:27   
4    delivered      2018-08-08 08:38:49  2018-08-08 08:55:23   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-10-04 19:55:00           2017-10-10 21:25:13   
1          2017-10-04 19:55:00           2017-10-10 21:25:13   
2          2017-10-04 19:55:00           2017-10-10 21:25:13   
3          2018-07-26 14:31:00           2018-08-07 15:27:45   
4          2018-08-08 13:50:00           2018-08-17 18:06:29   

  order_estimated_delivery_date delivery_status  order_item_id  ...   price  \
0           2017-10-18 00:00:00         on time              1  ...   29.99   
1           2017-10-18 00:00:00         on time              1  ...   29.99   
2           2017-10-18 00:00:00         on time              1  ...   29.99   
3           2018-08-13 00:00:00         on time              1  ...  118.70   
4           2018-09-04 00:00:00         on time              1  ...  159.90   

  freight_value payment_sequential  payment_type  payment_installments  \
0          8.72                  1   credit_card                     1   
1          8.72                  3       voucher                     1   
2          8.72                  2       voucher                     1   
3         22.76                  1        boleto                     1   
4         19.22                  1   credit_card                     3   

   payment_value                customer_unique_id  customer_zip_code_prefix  \
0          18.12  7c396fd4830fd04220f754e42b4e5bff                      3149   
1           2.00  7c396fd4830fd04220f754e42b4e5bff                      3149   
2          18.59  7c396fd4830fd04220f754e42b4e5bff                      3149   
3         141.46  af07308b275d755c9edb36a90c618231                     47813   
4         179.12  3a653a41f6f9fc3d2a113cf8398680e8                     75265   

   customer_city customer_state  
0      sao paulo             SP  
1      sao paulo             SP  
2      sao paulo             SP  
3      barreiras             BA  
4     vianopolis             GO  

[5 rows x 23 columns]

In [255]:
payment_report = customer_payment.groupby(by="order_id").agg({
    "customer_id": "nunique",
    "payment_value": "sum",
    "price": "sum",
    "freight_value": "sum",
    "order_purchase_timestamp": "max"
}).reset_index()

payment_report['outstanding_payment'] = payment_report['price'] + payment_report['freight_value'] - payment_report['payment_value']

payment_report.head()

order_id  customer_id  payment_value   price  \
0  00010242fe8c5a6d1ba2dd792cb16214            1          72.19   58.90   
1  00018f77f2f0320c557190d7a144bdd3            1         259.83  239.90   
2  000229ec398224ef6ca0657da4fc703e            1         216.87  199.00   
3  00024acbcdf0a6daa1e931b038114c75            1          25.78   12.99   
4  00042b26cf59d7ce69dfabb4e55b4fd9            1         218.04  199.90   

   freight_value order_purchase_timestamp  outstanding_payment  
0          13.29      2017-09-13 08:59:02         0.000000e+00  
1          19.93      2017-04-26 10:53:06         0.000000e+00  
2          17.87      2018-01-14 14:33:31         0.000000e+00  
3          12.79      2018-08-08 10:00:35         0.000000e+00  
4          18.14      2017-02-04 13:57:51         2.842171e-14

#### Customer yang belum melunasi pembayaran

In [256]:
payment_report[payment_report['outstanding_payment'] >= 1]

order_id  customer_id  payment_value   price  \
35     0016dfedd97fc2950e388d2971d718c7            1          70.55   99.50   
158    0071ee2429bc1efdc43aa3e073a5290e            1         192.44  359.96   
212    009ac365164f8e06f59d18a08045f6c4            1          32.00  101.40   
251    00b4a910f64f24dbcac04fe54088a443            1          50.59   67.98   
266    00bd50cdd31bd22e9081e6e2d5b3577b            1          85.80  234.00   
...                                 ...          ...            ...     ...   
96259  ff7400d904161b62b6e830b3988f5cbd            1         154.96  233.98   
96311  ff978de32e717acd3b5abe1fb069d2b6            1          49.14  111.96   
96322  ffa1dd97810de91a03abd7bd76d2fed1            1         455.31  869.98   
96326  ffa39020fe7c8a3e907320e1bec4b985            1          71.14  113.98   
96378  ffc730a0615d28ec19f9cad02cb41442            1          27.39   39.98   

       freight_value order_purchase_timestamp  outstanding_payment  
35             41.60      2017-04-28 19:54:40                70.55  
158            24.92      2018-01-22 11:08:15               192.44  
212            90.60      2017-11-23 00:03:52               160.00  
251            33.20      2017-01-18 17:55:34                50.59  
266            23.40      2018-01-14 22:31:54               171.60  
...              ...                      ...                  ...  
96259          75.94      2018-03-27 15:37:03               154.96  
96311          84.60      2017-12-12 09:50:42               147.42  
96322          40.64      2017-05-22 22:35:09               455.31  
96326          28.30      2017-10-31 18:34:34                71.14  
96378          14.80      2018-07-25 14:03:13                27.39  

[2662 rows x 7 columns]

#### Customer yang telah melunasi pembayaran

In [257]:
payment_report[payment_report['outstanding_payment'] < 1]

order_id  customer_id  payment_value   price  \
0      00010242fe8c5a6d1ba2dd792cb16214            1          72.19   58.90   
1      00018f77f2f0320c557190d7a144bdd3            1         259.83  239.90   
2      000229ec398224ef6ca0657da4fc703e            1         216.87  199.00   
3      00024acbcdf0a6daa1e931b038114c75            1          25.78   12.99   
4      00042b26cf59d7ce69dfabb4e55b4fd9            1         218.04  199.90   
...                                 ...          ...            ...     ...   
96453  fffc94f6ce00a00581880bf54a75a037            1         343.40  299.99   
96454  fffcd46ef2263f404302a634eb57f7eb            1         386.53  350.00   
96455  fffce4705a9662cd70adb13d4a31832d            1         116.85   99.90   
96456  fffe18544ffabc95dfada21779c9644f            1          64.71   55.99   
96457  fffe41c64501cc87c801fd61db3f6244            1          55.79   43.00   

       freight_value order_purchase_timestamp  outstanding_payment  
0              13.29      2017-09-13 08:59:02         0.000000e+00  
1              19.93      2017-04-26 10:53:06         0.000000e+00  
2              17.87      2018-01-14 14:33:31         0.000000e+00  
3              12.79      2018-08-08 10:00:35         0.000000e+00  
4              18.14      2017-02-04 13:57:51         2.842171e-14  
...              ...                      ...                  ...  
96453          43.41      2018-04-23 13:57:06         0.000000e+00  
96454          36.53      2018-07-14 10:26:46         0.000000e+00  
96455          16.95      2017-10-23 17:07:56         1.421085e-14  
96456           8.72      2017-08-14 23:02:59         1.421085e-14  
96457          12.79      2018-06-09 17:00:18         0.000000e+00  

[93796 rows x 7 columns]

#### Sebaran customer berdasarkan state dan kota

In [258]:
sellers_df.groupby(by=["seller_state", "seller_city"]).agg({
  "seller_id": "nunique"
}).sort_values(ascending=False, by="seller_id")

seller_id
seller_state seller_city                      
SP           sao paulo                     694
PR           curitiba                      124
RJ           rio de janeiro                 93
MG           belo horizonte                 66
SP           ribeirao preto                 52
...                                        ...
RJ           tres rios                       1
             saquarema                       1
             sao pedro da aldeia             1
             santo antonio de padua          1
SC           camboriu                        1

[636 rows x 1 columns]

**Insight:**
- Data revenue pelanggan berada pada periode tahun 2016 - 2018 selama 12 bulan, namun tidak setip tahun memiliki periode bulan yang sama, seperti pada tahun 2016, hanya terdapat data penjualan bulan 9, 10, dan 12. Hal ini berarti data yang diperoleh dimulai dari bulan 9 tahun 2016 sampai dengan bulan 12 tahun 2018.
- Customer dan seller tersebar di beberapa state dan beberapa kota di dalam setiap state. Kota Sao Paulo dengan state SP menjadi kota penyumbang jumlah customer dan seller terbanyak
- Setiap orderan memiliki rating pada rentang 1 - 5
- Delivery order dibagi menjadi 3 kategori, on time (tepat waktu), late (terlambat atau tidak tepat waktu), other (kategori tidak delivered)
- Beberapa pelanggan melunasi pembayaran dan beberapa tidak

## Visualization & Explanatory Analysis

### Pertanyaan 1: Bagaimana revenue penjualan selama setahun?

In [259]:
revenue = order_order_id.groupby(by=['purchase_year', 'purchase_month', 'purchase_day']).agg({
  "price": "sum"
}).reset_index()

revenue['purchase_year'] = revenue['purchase_year'].astype('str')
revenue['purchase_month'] = revenue['purchase_month'].astype('str')
revenue['purchase_day'] = revenue['purchase_day'].astype('str')
revenue.head()

purchase_year purchase_month purchase_day    price
0          2016              9           15   134.97
1          2016             10            3   463.48
2          2016             10            4  8595.89
3          2016             10            5  6169.77
4          2016             10            6  5889.96

Karena data penjualan 2016 kurang lengkap, perlu dilakukan perbaikan data agar dapat digabungkan menjadi 1 plot

In [260]:
revenue_dict = revenue.to_dict('records')
revenue_dict

months = [i for i in range(1, 13)]
years = ['2016', '2017', '2018']
days = [i for i in range(1, 32)]


existing_records = {(item['purchase_year'], int(item['purchase_month']), int(item['purchase_day'])) for item in revenue_dict}

missing_records = [
    {'purchase_year': year, 'purchase_month': month, 'purchase_day': day, 'price': 0}
    for year in years for month in months for day in days
    if (year, month, day) not in existing_records
]

revenue_dict.extend(missing_records)

print(revenue_dict)
revenue_data = pd.DataFrame.from_dict(revenue_dict)
revenue_data['purchase_month'] = revenue_data['purchase_month'].astype(int)
revenue_data['purchase_day'] = revenue_data['purchase_day'].astype(int)
revenue_data = revenue_data.sort_values(by=['purchase_year','purchase_month', 'purchase_day'], ascending=True)
revenue_data['purchase_day'] = revenue_data['purchase_day'].astype(int)
revenue_data['purchase_month'] = revenue_data['purchase_month'].astype('str')
revenue_data

[{'purchase_year': '2016', 'purchase_month': '9', 'purchase_day': '15', 'price': 134.97}, {'purchase_year': '2016', 'purchase_month': '10', 'purchase_day': '3', 'price': 463.48}, {'purchase_year': '2016', 'purchase_month': '10', 'purchase_day': '4', 'price': 8595.89}, {'purchase_year': '2016', 'purchase_month': '10', 'purchase_day': '5', 'price': 6169.77}, {'purchase_year': '2016', 'purchase_month': '10', 'purchase_day': '6', 'price': 5889.96}, {'purchase_year': '2016', 'purchase_month': '10', 'purchase_day': '7', 'price': 6325.25}, {'purchase_year': '2016', 'purchase_month': '10', 'purchase_day': '8', 'price': 7692.88}, {'purchase_year': '2016', 'purchase_month': '10', 'purchase_day': '9', 'price': 2644.5}, {'purchase_year': '2016', 'purchase_month': '10', 'purchase_day': '10', 'price': 3159.57}, {'purchase_year': '2016', 'purchase_month': '12', 'purchase_day': '23', 'price': 10.9}, {'purchase_year': '2017', 'purchase_month': '1', 'purchase_day': '5', 'price': 396.9}, {'purchase_year'

purchase_year purchase_month  purchase_day  price
612           2016              1             1    0.0
613           2016              1             2    0.0
614           2016              1             3    0.0
615           2016              1             4    0.0
616           2016              1             5    0.0
...            ...            ...           ...    ...
1111          2018             12            27    0.0
1112          2018             12            28    0.0
1113          2018             12            29    0.0
1114          2018             12            30    0.0
1115          2018             12            31    0.0

[1116 rows x 4 columns]

In [261]:
revenue_data_by_year_month = revenue_data.groupby(by=['purchase_year', 'purchase_month']).agg({
  'price': 'sum'
}).reset_index()

revenue_data_by_year_month['purchase_month'] = revenue_data_by_year_month['purchase_month'].astype(int)
revenue_data_by_year_month = revenue_data_by_year_month.sort_values(by=['purchase_year', 'purchase_month'])
revenue_data_by_year_month['purchase_month'] = revenue_data_by_year_month['purchase_month'].astype('str')

fig = px.line(
    revenue_data_by_year_month,
    x="purchase_month",
    y="price",
    title='Revenue over years',
    color='purchase_year',
    markers=True,
    labels={
        "price": "Price (USD)",
        "purchase_year": "Year",
        "purchase_month": "Month"
    }
    )
fig.show()

### Pertanyaan 2: Bagaimana demografi customer yang dimiliki?

In [262]:
customers_geolocation_df = pd.merge(
  left=customers_df,
  right=geolocation_df,
  left_on="customer_zip_code_prefix",
  right_on="geolocation_zip_code_prefix"
)

customers_geolocation_df.head()

customer_id                customer_unique_id  \
0  06b8999e2fba1a1fbc88172c00ba8bc7  861eff4711a542e4b93843c6dd7febb0   
1  06b8999e2fba1a1fbc88172c00ba8bc7  861eff4711a542e4b93843c6dd7febb0   
2  06b8999e2fba1a1fbc88172c00ba8bc7  861eff4711a542e4b93843c6dd7febb0   
3  06b8999e2fba1a1fbc88172c00ba8bc7  861eff4711a542e4b93843c6dd7febb0   
4  06b8999e2fba1a1fbc88172c00ba8bc7  861eff4711a542e4b93843c6dd7febb0   

   customer_zip_code_prefix customer_city customer_state  \
0                     14409        franca             SP   
1                     14409        franca             SP   
2                     14409        franca             SP   
3                     14409        franca             SP   
4                     14409        franca             SP   

   geolocation_zip_code_prefix  geolocation_lat  geolocation_lng  \
0                        14409       -20.509897       -47.397866   
1                        14409       -20.497396       -47.399241   
2                        14409       -20.510459       -47.399553   
3                        14409       -20.480940       -47.394161   
4                        14409       -20.515413       -47.398194   

  geolocation_city geolocation_state  
0           franca                SP  
1           franca                SP  
2           franca                SP  
3           franca                SP  
4           franca                SP

In [263]:
customers_df.groupby(by="customer_state").agg({
  "customer_id": "nunique"
})

customer_id
customer_state             
AC                       81
AL                      413
AM                      148
AP                       68
BA                     3380
CE                     1336
DF                     2140
ES                     2033
GO                     2020
MA                      747
MG                    11635
MS                      715
MT                      907
PA                      975
PB                      536
PE                     1652
PI                      495
PR                     5045
RJ                    12852
RN                      485
RO                      253
RR                       46
RS                     5466
SC                     3637
SE                      350
SP                    41746
TO                      280

### Pertanyaan 3 : Product apa yang memiliki skor review terbaik dan terburuk?

In [264]:
product_review_score = orders_products_review.groupby(by="product_category_name").agg({
  "review_score": "sum"
}).reset_index()
product_review_score.head()

product_category_name  review_score
0  agro_industria_e_comercio         842.0
1                  alimentos        2076.0
2          alimentos_bebidas        1184.0
3                      artes         796.0
4         artes_e_artesanato          99.0

In [265]:
highest_review_scores = product_review_score.sort_values(ascending=True, by="review_score").tail(10)

lowest_review_scores = product_review_score.sort_values(ascending=True, by="review_score").head(10)

fig = make_subplots(
    rows=1, cols=2,
    subplot_titles=["10 Products with Highest Review Score", "10 Products with Lowest Review Score"],
    shared_yaxes=False  # Allow independent y-axes
)
fig.add_trace(
    go.Bar(
        y=highest_review_scores["product_category_name"],
        x=highest_review_scores["review_score"],
        text=highest_review_scores["review_score"],
        textposition="auto",
        orientation="h",
        marker_color="blue"
    ),
    row=1, col=1
)

fig.add_trace(
    go.Bar(
        y=lowest_review_scores["product_category_name"],
        x=lowest_review_scores["review_score"],
        text=lowest_review_scores["review_score"],
        textposition="auto",
        orientation="h",
        marker_color="red"
    ),
    row=1, col=2
)

fig.update_layout(
    height=600,
    width=1200,
    showlegend=False,
    xaxis_title="Review Score",
    xaxis2_title="Review Score",
    yaxis=dict(title="Product Category", automargin=True),
    yaxis2=dict(title="Product Category", automargin=True, side="right") 
)

fig.show()


### Berapa persentase pesanan diantar tepat waktu dan terlambat?

In [266]:
delivery_percentage = orders_df.groupby(by='delivery_status').agg({
  "order_id": "nunique"
}).sort_values(ascending=False, by="delivery_status").reset_index()
fig = px.pie(
  delivery_percentage,
  values="order_id",
  names="delivery_status",
  title="Persentase pesanan terlambat dan tepat waktu",
  width=600,
  height=600
  )
fig.update_traces(hole=.55)
fig.show()

### Bagaimana persentase pelanggan yang melunasi pembayaran order terhadap total pelanggan?

In [267]:
payment_report.head()

order_id  customer_id  payment_value   price  \
0  00010242fe8c5a6d1ba2dd792cb16214            1          72.19   58.90   
1  00018f77f2f0320c557190d7a144bdd3            1         259.83  239.90   
2  000229ec398224ef6ca0657da4fc703e            1         216.87  199.00   
3  00024acbcdf0a6daa1e931b038114c75            1          25.78   12.99   
4  00042b26cf59d7ce69dfabb4e55b4fd9            1         218.04  199.90   

   freight_value order_purchase_timestamp  outstanding_payment  
0          13.29      2017-09-13 08:59:02         0.000000e+00  
1          19.93      2017-04-26 10:53:06         0.000000e+00  
2          17.87      2018-01-14 14:33:31         0.000000e+00  
3          12.79      2018-08-08 10:00:35         0.000000e+00  
4          18.14      2017-02-04 13:57:51         2.842171e-14

In [268]:
pd.to_datetime(payment_report['order_purchase_timestamp']).dt.year

0        2017
1        2017
2        2018
3        2018
4        2017
         ... 
96453    2018
96454    2018
96455    2017
96456    2017
96457    2018
Name: order_purchase_timestamp, Length: 96458, dtype: int32

In [269]:
complete_count = payment_report[payment_report['outstanding_payment'] < 1].shape[0]
total_count = payment_report.shape[0]
incomplete_count = total_count - complete_count

data = {
  'name' : ['belum lunas', 'lunas'],
  'count' : [incomplete_count, complete_count]
}

df = pd.DataFrame.from_dict(data)

fig = px.pie(
  df,
  values="count",
  names="name",
  title="Persentase pelanggan melunasi terhadap total pelanggan",
  width=600,
  height=600
  )
fig.update_traces(textinfo='percent+label', texttemplate='%{percent:.1%}', hole=.55)
fig.show()


### Bagaimana demografi seller yang dimiliki?

In [270]:
sellers_df.groupby(by="seller_state").agg({
  "seller_id": "nunique"
})

seller_id
seller_state           
AC                    1
AM                    1
BA                   19
CE                   13
DF                   30
ES                   23
GO                   40
MA                    1
MG                  244
MS                    5
MT                    4
PA                    1
PB                    6
PE                    9
PI                    1
PR                  349
RJ                  171
RN                    5
RO                    2
RS                  129
SC                  190
SE                    2
SP                 1849

**Insight:**
- Revenue penjualan relative menaik pada tahun 2017. Pada tahun 2016 revenue penjualan cenderung menurun, namun hal ini hanya dibuktikan oleh 3 bulan penjualan yaitu bulan 10, 11, dan 12. Pada awal tahun 2018, revenue berada jauh di atas 2 tahun sebelumnya, dengan rata - rata penjualan setiap bulannya secara berturut - turut lebih tinggi dari setiap bulan pada 2 tahun sebelumnya, namun demikian pada bulan 9 hingga bulan 12 tahun 2018, penjualan jatuh ke angka 0, hal ini berarti tidak ada penjualan sama sekali pada periode tersebut
- Produk yang paling diminati oleh customer berdasarkan review skor adalah cama_mesa_banho, sementara itu produk yang paling sedikit diminati oleh customer adalah seguros_e_servicos
- Terdapat lebih banyak pesanan yang diantar tepat waktu (on time)
 yaitu sekitar 91.9% dibandingkan pesanan yang terlambat diantar (late) 8.1%, diikuti dengan status pengiriman lain (other), yaitu sebesar 0.00622%, status ini merupakan pengiriman yang sedang berada di perjalanan. Hal ini membuktikan bahwa sirkulasi pembelian dan pengiriman mengalami sinkronasi yang baik
- Terdapat lebih banyak karyawan yang melunasi orderan yaitu sebesar 97.2%, baik dengan menyicil atau dengan membayar lunas dibandingkan dengan yang belum melunasi, yaitu sebesar 2.8%. Hal ini berarti koitment beli customer cukup tinggi terhadap orderan

## Analisis Lanjutan (Opsional)

### RFM (Recency, Frequency, Monetary)  Analysis

In [271]:
order_customer_df = pd.merge(
  left=orders_df,
  right=customers_df,
  left_on="customer_id",
  right_on="customer_id"
)

order_customer_payment_df = pd.merge(
  left=order_customer_df,
  right=order_payments_df,
  left_on="order_id",
  right_on="order_id"
)

order_customer_payment_df.head()

order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
2  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
3  53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
4  47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   

  order_status order_purchase_timestamp    order_approved_at  \
0    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
1    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
2    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
3    delivered      2018-07-24 20:41:37  2018-07-26 03:24:27   
4    delivered      2018-08-08 08:38:49  2018-08-08 08:55:23   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-10-04 19:55:00           2017-10-10 21:25:13   
1          2017-10-04 19:55:00           2017-10-10 21:25:13   
2          2017-10-04 19:55:00           2017-10-10 21:25:13   
3          2018-07-26 14:31:00           2018-08-07 15:27:45   
4          2018-08-08 13:50:00           2018-08-17 18:06:29   

  order_estimated_delivery_date delivery_status  \
0           2017-10-18 00:00:00         on time   
1           2017-10-18 00:00:00         on time   
2           2017-10-18 00:00:00         on time   
3           2018-08-13 00:00:00         on time   
4           2018-09-04 00:00:00         on time   

                 customer_unique_id  customer_zip_code_prefix customer_city  \
0  7c396fd4830fd04220f754e42b4e5bff                      3149     sao paulo   
1  7c396fd4830fd04220f754e42b4e5bff                      3149     sao paulo   
2  7c396fd4830fd04220f754e42b4e5bff                      3149     sao paulo   
3  af07308b275d755c9edb36a90c618231                     47813     barreiras   
4  3a653a41f6f9fc3d2a113cf8398680e8                     75265    vianopolis   

  customer_state  payment_sequential payment_type  payment_installments  \
0             SP                   1  credit_card                     1   
1             SP                   3      voucher                     1   
2             SP                   2      voucher                     1   
3             BA                   1       boleto                     1   
4             GO                   1  credit_card                     3   

   payment_value  
0          18.12  
1           2.00  
2          18.59  
3         141.46  
4         179.12

- Kapan terakhir kali pelanggan melakukan transaksi
- Berapa jumlah transaksi pelanggan dalam periode tertentu?
- Berapa total pengeluaran pelanggan pada periode tertentu?

In [272]:
customer_rfm_df = order_customer_payment_df.groupby(by='customer_id').agg({
  'order_purchase_timestamp': 'max',
  'order_id': 'nunique',
  'payment_value': 'sum'
}).reset_index()

customer_rfm_df['order_purchase_timestamp'] = pd.to_datetime(customer_rfm_df['order_purchase_timestamp']).dt.date
last_date = pd.to_datetime(customer_rfm_df['order_purchase_timestamp']).dt.date.max()

customer_rfm_df['recency'] = customer_rfm_df['order_purchase_timestamp'].apply(lambda x: (last_date - x).days)

customer_rfm_df.drop('order_purchase_timestamp', axis=1, inplace=True)
customer_rfm_df.columns = ['customer_id', 'frequency', 'monetary', 'recency']
customer_rfm_df.head()

customer_id  frequency  monetary  recency
0  00012a2ce6f8dcda20d059ce98491703          1    114.74      288
1  000161a058600d5901f007fab4c27140          1     67.41      409
2  0001fd6190edaaf884bcaf3d49edf079          1    195.42      547
3  0002414f95344307404f0ace7a26f1d5          1    179.35      378
4  000379cdec625522490c315e70c7a9fb          1    107.01      149

In [273]:
customer_rfm_df.sort_values(by='monetary', ascending=False)

customer_id  frequency  monetary  recency
8291   1617b1357756262bfa56ab541c47bc16          1  13664.08      334
89257  ec5b2ba62e574342386871631fafd3fc          1   7274.88       45
75226  c6e2731c5b391845f6800c97401a43a9          1   6929.31      563
92292  f48d464a0baaea338cb25f816991ab1f          1   6922.21       35
24061  3fd6777bbce08a352fddd04e4a7cc8f6          1   6726.66      462
...                                 ...        ...       ...      ...
79690  d2c63ad286e3ca9dd69218008d61ff81          1     11.62      601
1690   046f890135acc703faff4c1fc0c2d73c          1     11.56       68
53523  8e4bd65db637116b6b68109e4df21b84          1     10.89       69
9176   184e8e8e48937145eb96c721ef1f0747          1     10.07      350
63163  a790343ca6f3fee08112d678b43aa7c5          1      9.59        4

[96458 rows x 4 columns]

In [274]:
customer_rfm_df.sort_values(by='recency', ascending=False)

customer_id  frequency  monetary  recency
45176  7812fcebfc5e8065d31e1bb5f0017dae          1     40.95      695
47744  7ec40b22510fdbea1b08921dd39e63d8          1     39.09      695
87243  e6f959bf384d1d53b6d68826699bba12          1    154.57      695
20112  355077684019f7f60a031656bd7262b8          1     45.46      695
41914  6f989332712d3222b6571b1cf5b835ce          1     53.73      695
...                                 ...        ...       ...      ...
86189  e450a297a7bc6839ceb0cf1a2377fa02          1     73.10        0
31658  54365416b7ef5599f54a6c7821d5d290          1    106.95        0
86862  e60df9449653a95af4549bbfcb18a6eb          1    510.96        0
41373  6e353700bc7bcdf6ebc15d6de16d7002          1     61.29        0
51748  898b7fee99c4e42170ab69ba59be0a8b          1     93.75        0

[96458 rows x 4 columns]

## Conclusion

- Customer dengan total pengeluaran terbanyak adalah customer dengan id : 1617b1357756262bfa56ab541c47bc16, yaitu sebesar $13,664.08. Customer ini mungkin cocok untuk diberi reward melalui hadiah berupa voucher belanja atau diskon agar tetap menjaga loyalitas
- Customer dengan tanggal pembelian terlama adalah customer dengan id : 7812fcebfc5e8065d31e1bb5f0017dae, sementara itu terdapat beberapa customer yang berbelanja pada hari yang sama berdasarkan periode dataset (2016 - 2018).
- Frequency customer masing - masing adalah sebesar 1, hal ini bisa menjadi indikasi bahwa customer tidak lagi membeli barang di tempat yang sama. Jika customer id adalah identifier unique untuk tiap customer yang membeli barang, maka sangat disayangkan jika customer tidak kembali lagi setelah membeli barang, mungkin dapat disusun strategi yang lebih baik untuk menarik minat sekaligus menjaga kepercayaan customer agar berbelanja kembali

In [275]:
revenue_data.to_csv('./dashboard/revenue.csv', index=False)
orders_products_review.to_csv('./dashboard/order_product_review.csv', index=False)
payment_report.to_csv('./dashboard/payment_report.csv', index=False)
customers_df.to_csv('./dashboard/customers.csv')